File: Apps/Android/app/src/main/kotlin/com/opencapture/openpocketcine/diagnostics/FeedIncidentRuntime.kt
package com.opencapture.openpocketcine.diagnostics

import java.io.File
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Shell owner for feed-incident persistence. Recorder stays in memory; this
 * type writes on a dedicated utility thread. Never call persist from receive,
 * decode, display, or fatal-signal handlers.
 */
internal object FeedIncidentRuntime {
    private val writer =
        Executors.newSingleThreadExecutor { thread ->
            Thread(thread, "opc.feed-incident").apply { isDaemon = true }
        }
    private val recorder = FeedIncidentRecorder()
    private var store: FeedIncidentStore? = null
    @Volatile private var latestSnapshot: FeedIncidentSnapshot? = null
    private val snapshotScheduled = AtomicBoolean(false)
    private var sessionContext: FeedIncidentSessionContext? = null
    private var lastSummaryCheckpoint = 0.0

    fun install(directory: File) {
        writer.execute {
            val created = FeedIncidentStore(File(directory, "incidents"))
            store = created
            created.markInterrupted()
            FeedSessionSummaryStore.markInterrupted(created.directory)
        }
    }

    fun deleteStoredReports(completion: (Boolean) -> Unit) {
        writer.execute {
            val deleted = runCatching {
                val saved = store ?: return@runCatching false
                val incidentsDeleted = saved.deleteAll()
                val summariesDeleted = FeedSessionSummaryStore.deleteAll(saved.directory)
                incidentsDeleted && summariesDeleted
            }.getOrDefault(false)
            android.os.Handler(android.os.Looper.getMainLooper()).post { completion(deleted) }
        }
    }

    fun beginSession(context: FeedIncidentSessionContext) {
        writer.execute {
            persist(recorder.beginSession(context))
            sessionContext = context
            lastSummaryCheckpoint = 0.0
            persistSessionSummary("live")
        }
    }

    fun endSession(nowSeconds: Double) {
        writer.execute {
            persist(recorder.endSession(nowSeconds))
            persistSessionSummary(summaryOutcome(recorder.incidentCount, recorder.healthyExposure))
            sessionContext = null
            ReliabilityReporting.enqueueFinalizedFromSpool()
        }
    }

    fun ingestSnapshot(snapshot: FeedIncidentSnapshot) {
        latestSnapshot = snapshot
        if (snapshotScheduled.compareAndSet(false, true)) {
            writer.execute { drainSnapshot() }
        }
    }

    fun recordBreadcrumb(breadcrumb: FeedIncidentBreadcrumb) {
        writer.execute {
            recorder.recordBreadcrumb(breadcrumb)
            ReliabilityReporting.noteBreadcrumb(breadcrumb)
        }
    }

    fun recordRepair(repair: FeedRepairRecord) {
        writer.execute {
            persist(recorder.recordRepair(repair))
            ReliabilityReporting.noteRepair(repair)
        }
    }

    fun noteTestSource(source: FeedIncidentTestSource) {
        writer.execute {
            recorder.noteTestSource(source)
            val context = sessionContext
            if (context != null && source.rank > context.testSource.rank) {
                context.testSource = source
            }
        }
    }

    fun noteDecoderGeneration(generation: Int) {
        writer.execute { recorder.noteDecoderGeneration(generation) }
    }

    fun noteSocketGeneration(generation: Int) {
        writer.execute { recorder.noteSocketGeneration(generation) }
    }

    fun noteExhausted(nowSeconds: Double) {
        writer.execute { persist(recorder.noteExhausted(nowSeconds)) }
    }

    fun noteUnexpectedDisconnect(nowSeconds: Double) {
        writer.execute { persist(recorder.noteUnexpectedDisconnect(nowSeconds)) }
    }

    fun exportText(): String {
        val future = writer.submit<String> { store?.exportText().orEmpty() }
        return runCatching { future.get(2, java.util.concurrent.TimeUnit.SECONDS) }.getOrDefault("")
    }

    fun exportExtras(): List<Pair<String, String>> {
        val future =
            writer.submit<List<Pair<String, String>>> { store?.exportExtras().orEmpty() }
        return runCatching { future.get(2, java.util.concurrent.TimeUnit.SECONDS) }.getOrDefault(emptyList())
    }

    /** Manual-report extras: typed incidents plus compact session summaries. */
    fun reportExtras(): List<Pair<String, String>> {
        val extras = exportExtras().toMutableList()
        val summaries = loadSessionSummaries().sortedByDescending { it.recordedAtMs }
        summaries.firstOrNull()?.let { newest ->
            extras += "session-summary.json" to PrivacyRedactor.redact(FeedSessionSummaryStore.encodeForReport(newest))
        }
        for (summary in summaries.drop(1).take(4)) {
            extras +=
                "session-${summary.sessionID}.json" to
                    PrivacyRedactor.redact(FeedSessionSummaryStore.encodeForReport(summary))
        }
        return extras
    }

    fun loadFinalized(): List<FeedIncidentBundle> {
        val future =
            writer.submit<List<FeedIncidentBundle>> {
                store?.loadAll()?.filter { ReliabilityReporting.isFinalized(it) }.orEmpty()
            }
        return runCatching { future.get(2, java.util.concurrent.TimeUnit.SECONDS) }.getOrDefault(emptyList())
    }

    fun loadSessionSummaries(): List<FeedIncidentSessionSummary> {
        val future =
            writer.submit<List<FeedIncidentSessionSummary>> {
                val root = store?.directory ?: return@submit emptyList()
                FeedSessionSummaryStore.load(root).filter { it.outcome != "live" }
            }
        return runCatching { future.get(2, java.util.concurrent.TimeUnit.SECONDS) }.getOrDefault(emptyList())
    }

    fun summaryOutcome(incidentCount: Int, exposure: Double): String {
        if (incidentCount > 0) return "ended"
        if (exposure > 0.0) return "healthy"
        return "no-exposure"
    }

    private fun drainSnapshot() {
        val snapshot = latestSnapshot
        latestSnapshot = null
        snapshotScheduled.set(false)
        if (snapshot != null) {
            persist(recorder.recordSnapshot(snapshot))
            if (snapshot.monotonicNow - lastSummaryCheckpoint >= 30.0) {
                lastSummaryCheckpoint = snapshot.monotonicNow
                persistSessionSummary("live")
            }
        }
    }

    private fun persistSessionSummary(outcome: String) {
        val context = sessionContext ?: return
        val summary =
            FeedIncidentSessionSummary(
                sessionID = context.sessionId,
                healthyExposureSeconds = recorder.healthyExposure,
                incidentCount = recorder.incidentCount,
                outcome = outcome,
                sourceRevision = context.sourceRevision,
                appVersion = context.appVersion,
                appBuild = context.appBuild,
                testSource = context.testSource,
                buildIdentity = context.buildIdentity,
            )
        val root = store?.directory ?: return
        FeedSessionSummaryStore.persist(summary, root)
        if (summary.outcome != "live") ReliabilityReporting.noteSessionSummary(summary)
    }

    private fun persist(job: FeedIncidentPersistenceJob?) {
        if (job == null) return
        runCatching { store?.persist(job) }
        if (ReliabilityReporting.isFinalized(job.bundle)) {
            ReliabilityReporting.enqueueFinalized(job.bundle)
        }
    }
}

File: ios/OpenPocketCine/FeedIncidentRuntime.swift
import Foundation
import OpenPocketViewCore

/// Shell owner for feed-incident persistence. Core records in memory; this type
/// writes on a utility queue. Never call from a receive, decode, display, or
/// fatal-signal handler except via the ingest methods, which return immediately.
///
/// Coordinator hooks (camera session / decoder / chrome — not wired here):
/// - `install()` once at diagnostic boot (marks leftover open incidents interrupted)
/// - `beginSession(_:)` when a live session identity is known
/// - `endSession(now:)` on explicit disconnect
/// - `ingestSnapshot(_:)` at 1 Hz with allowlisted stage counters
/// - `recordBreadcrumb(_:)` Settings enter/exit, assist, scene, surface, decoder, path, command category
/// - `recordRepair(_:)` requested / blocked / locallySent / peerResponse / pictureRestored
/// - `noteDecoderGeneration(_:)` / `noteSocketGeneration(_:)`
/// - `noteExhausted(now:)` when the repair owner spends its budget
/// - `noteUnexpectedDisconnect(now:)` for BLE/path loss before ages exist
/// - `reportExtras()` from Share Diagnostics (journal-independent)
struct FeedIncidentSessionSummary: Equatable, Codable {
    var sessionID: String
    var healthyExposureSeconds: Double
    var incidentCount: Int
    var outcome: String
    var sourceRevision: String
    var recordedAt: Date? = Date()
    var appVersion: String? = nil
    var appBuild: String? = nil
    var testSource: FeedIncidentTestSource? = nil
    var buildIdentity: String? = nil
}

enum FeedIncidentRuntime {
    private static let queue = DispatchQueue(label: "opv.feed-incident", qos: .utility)
    private static let gate = NSLock()
    private static let extrasLock = NSLock()
    private static var recorder = FeedIncidentRecorder()
    private static var store: FeedIncidentStore?
    private static var latestSnapshot: FeedIncidentSnapshot?
    private static var snapshotScheduled = false
    private static var pendingDecoderGeneration: Int?
    private static var pendingSocketGeneration: Int?
    private static var sessionContext: FeedIncidentSessionContext?
    private static var lastSummaryCheckpoint: TimeInterval = 0
    private static var extrasCache: [(name: String, body: String)] = []
    /// Queue-confined export of the incident bundles on disk; nil reloads them.
    private static var storeExtras: [(name: String, body: String)]?

    static func install() {
        queue.async {
            _ = preparedStore()
            _ = try? store?.markInterrupted(now: Date())
            if let root = store?.directory { FeedSessionSummaryStore.markInterrupted(root: root) }
            ReliabilityReporting.setFinalizedSpoolLoader {
                let dir = DiagnosticCenter.diagnosticsDirectory?
                    .appendingPathComponent("incidents", isDirectory: true)
                guard let dir else { return [] }
                return FeedIncidentStore(directory: dir).loadAll().filter {
                    ReliabilityReporting.isFinalized($0)
                }
            }
            ReliabilityReporting.setSessionSummaryLoader {
                guard
                    let root = DiagnosticCenter.diagnosticsDirectory?
                        .appendingPathComponent("incidents", isDirectory: true)
                else { return [] }
                return FeedSessionSummaryStore.load(root: root).filter { $0.outcome != "live" }
            }
            refreshExtrasCache()
            ReliabilityReporting.enqueueFinalizedFromSpool()
        }
    }

    static func beginSession(_ context: FeedIncidentSessionContext) {
        queue.async {
            persist(recorder.beginSession(context))
            ReliabilityReporting.enqueueFinalizedFromSpool()
            sessionContext = context
            lastSummaryCheckpoint = 0
            persistSessionSummary(outcome: "live")
        }
    }

    static func endSession(now: TimeInterval) {
        queue.async {
            persist(recorder.endSession(now: now))
            persistSessionSummary(
                outcome: summaryOutcome(
                    incidentCount: recorder.incidentCount,
                    exposure: recorder.healthyExposureSeconds))
            ReliabilityReporting.enqueueFinalizedFromSpool()
            sessionContext = nil
        }
    }

    /// Latest-wins 1 Hz ingest. Does not enqueue one disk write per snapshot.
    static func ingestSnapshot(_ snapshot: FeedIncidentSnapshot) {
        gate.lock()
        latestSnapshot = snapshot
        let schedule = !snapshotScheduled
        snapshotScheduled = true
        gate.unlock()
        if schedule {
            queue.async { drainSnapshot() }
        }
    }

    static func recordBreadcrumb(_ breadcrumb: FeedIncidentBreadcrumb) {
        queue.async {
            recorder.recordBreadcrumb(breadcrumb)
            ReliabilityReporting.noteBreadcrumb(breadcrumb)
        }
    }

    static func recordRepair(_ repair: FeedRepairRecord) {
        queue.async {
            persist(recorder.recordRepair(repair))
            ReliabilityReporting.noteRepair(repair)
        }
    }

    static func noteTestSource(_ source: FeedIncidentTestSource) {
        queue.async {
            recorder.noteTestSource(source)
            if var context = sessionContext, source.rank > context.testSource.rank {
                context.testSource = source
                sessionContext = context
            }
        }
    }

    static func noteDecoderGeneration(_ generation: Int) {
        scheduleGeneration { pendingDecoderGeneration = generation }
    }

    static func noteSocketGeneration(_ generation: Int) {
        scheduleGeneration { pendingSocketGeneration = generation }
    }

    static func noteExhausted(now: TimeInterval) {
        queue.async { persist(recorder.noteExhausted(now: now)) }
    }

    static func noteUnexpectedDisconnect(now: TimeInterval) {
        queue.async { persist(recorder.noteUnexpectedDisconnect(now: now)) }
    }

    /// Memory snapshot of extras. Does not wait on disk.
    static func reportExtras() -> [(name: String, body: String)] {
        extrasLock.lock()
        let copy = extrasCache
        extrasLock.unlock()
        return copy
    }

    /// Typed incident export. Completes off the live path; never writes on MainActor.
    static func exportVendorBundles(completion: @escaping (URL?) -> Void) {
        queue.async {
            let extras = preparedStore()?.exportExtras() ?? []
            let body =
                extras.first { $0.name == "incidents.txt" }?.body
                ?? "OpenPocketCine feed incidents\ncount=0"
            let url = DiagnosticCenter.diagnosticsDirectory?
                .appendingPathComponent("incidents-share.txt")
            if let url {
                try? body.write(to: url, atomically: true, encoding: .utf8)
            }
            DispatchQueue.main.async { completion(url) }
        }
    }

    /// Removes local incident bundles. Does not purge the SDK cache.
    static func deleteStoredIncidents(completion: @escaping () -> Void) {
        queue.async {
            try? preparedStore()?.deleteAll()
            if let root = store?.directory { FeedSessionSummaryStore.deleteAll(root: root) }
            if let dir = preparedStore()?.directory {
                try? FileManager.default.removeItem(
                    at: dir.appendingPathComponent("session-summary.json"))
            }
            storeExtras = nil
            extrasLock.lock()
            extrasCache = []
            extrasLock.unlock()
            DispatchQueue.main.async { completion() }
        }
    }

    static func summaryOutcome(incidentCount: Int, exposure: TimeInterval) -> String {
        if incidentCount > 0 { return "ended" }
        if exposure > 0 { return "healthy" }
        return "no-exposure"
    }

    static func seedExtrasCacheForTests(_ extras: [(name: String, body: String)]) {
        extrasLock.lock()
        extrasCache = extras
        extrasLock.unlock()
    }

    private static func scheduleGeneration(_ update: () -> Void) {
        gate.lock()
        update()
        let schedule = !snapshotScheduled
        snapshotScheduled = true
        gate.unlock()
        if schedule {
            queue.async { drainSnapshot() }
        }
    }

    private static func drainSnapshot() {
        gate.lock()
        let snapshot = latestSnapshot
        latestSnapshot = nil
        let decoderGen = pendingDecoderGeneration
        pendingDecoderGeneration = nil
        let socketGen = pendingSocketGeneration
        pendingSocketGeneration = nil
        snapshotScheduled = false
        gate.unlock()
        if let decoderGen { recorder.noteDecoderGeneration(decoderGen) }
        if let socketGen { recorder.noteSocketGeneration(socketGen) }
        if let snapshot {
            persist(recorder.recordSnapshot(snapshot))
            if snapshot.monotonicNow - lastSummaryCheckpoint >= 30 {
                lastSummaryCheckpoint = snapshot.monotonicNow
                persistSessionSummary(outcome: "live")
            }
        }
    }

    private static func persist(_ job: FeedIncidentPersistenceJob?) {
        guard let job = job, let store = preparedStore() else { return }
        try? store.persist(job)
        storeExtras = nil
        refreshExtrasCache()
    }

    private static func persistSessionSummary(outcome: String) {
        guard let summary = currentSessionSummary(outcome: outcome) else { return }
        guard let dir = preparedStore()?.directory else { return }
        FeedSessionSummaryStore.persist(summary, root: dir)
        // Summaries live outside the incident bundles, so the 30 s live checkpoint
        // does not reload and re-decode every stored incident.
        refreshExtrasCache()
        if summary.outcome != "live" { ReliabilityReporting.noteSessionSummary(summary) }
    }

    private static func currentSessionSummary(outcome: String? = nil)
        -> FeedIncidentSessionSummary?
    {
        guard let sessionContext else { return nil }
        return FeedIncidentSessionSummary(
            sessionID: sessionContext.sessionID,
            healthyExposureSeconds: recorder.healthyExposureSeconds,
            incidentCount: recorder.incidentCount,
            outcome: outcome
                ?? recorder.openHeader?.outcome.rawValue
                ?? summaryOutcome(
                    incidentCount: recorder.incidentCount,
                    exposure: recorder.healthyExposureSeconds),
            sourceRevision: sessionContext.sourceRevision,
            appVersion: sessionContext.appVersion, appBuild: sessionContext.appBuild,
            testSource: sessionContext.testSource,
            buildIdentity: sessionContext.buildIdentity)
    }

    private static func refreshExtrasCache() {
        if storeExtras == nil { storeExtras = preparedStore()?.exportExtras() }
        var extras = storeExtras ?? []
        if let summary = currentSessionSummary(),
            let data = try? FeedIncidentJSON.encoder().encode(summary),
            let body = String(data: data, encoding: .utf8)
        {
            extras.append(("session-summary.json", body))
        }
        extrasLock.lock()
        extrasCache = extras
        extrasLock.unlock()
    }

    private static func preparedStore() -> FeedIncidentStore? {
        if let store = store { return store }
        guard let dir = DiagnosticCenter.diagnosticsDirectory else { return nil }
        let incidents = dir.appendingPathComponent("incidents", isDirectory: true)
        let created = FeedIncidentStore(directory: incidents)
        store = created
        return created
    }
}
